casper/src/test/scala/coop/rchain/casper/wip/MergeRejectionSpec.scala

```diff
@@ -0,0 +1,147 @@
+    |  @"READ"!(*retCh) |
+    |  for(@s <- retCh) {
+    |    out!(("Read st:", s))
+    |  }
+    |}
+    |""".stripMargin
+
+  def testCase[F[_]: Sync](
+      nodes: IndexedSeq[TestNode[F]]
+  )(addDeploysNode2: List[Signed[DeployData]] => F[List[BlockMessage]]) = {
+    val deployST = deploy(rhoST)
+
+    for {
+      // Node 0 creates initial contracts
+      blockST <- nodes(0).addBlock(deployST)
+      // Node 1 & 2 accepts initial contracts
+      _ <- nodes(1).processBlock(blockST)
+      _ <- nodes(2).processBlock(blockST)
+
+      // Node 1 change
+      deployNil1a = deploy("Nil")
+      deploySet1  = deploy(rhoSet(333))
+      deployNil1b = deploy("Nil")
+      deploysNode1 = ListMap(
+        "N1_nila" -> deployNil1a,
+        "N1_set"  -> deploySet1,
+        "N1_nilb" -> deployNil1b
+      )
+      blocksNode1 <- deploysNode1.values.toList.traverse(nodes(1).addBlock(_))
+
+      // Node 2 change & read
+      deploySet2   = deploy(rhoSet(444), defaultSec2)
+      deployRead2  = deploy(rhoRead, defaultSec2)
+      deploysNode2 = ListMap("N2_set" -> deploySet2, "N2_read" -> deployRead2)
+      // Add deploys in separate blocks or in the same block
+      blocksNode2 <- addDeploysNode2(deploysNode2.values.toList)
+
+      // All blocks
+      blocks      = blocksNode1 ++ blocksNode2
+      blockHashes = blocks.map(_.blockHash)
+
+      // Node 0 accepts all blocks from node 1 and 2
+      _ <- blocks.traverse(nodes(0).processBlock)
+
+      // Node 0 creates merged block
+      deployNil   = deploy("Nil")
+      blockMerged <- nodes(0).addBlock(deployNil)
+      // Node 0 blocks
+      node0Blocks <- blockHashes.traverse(nodes(0).contains)
+
+      // Node 0 merged parents
+      mergedParents = blockHashes.map(blockMerged.header.parentsHashList.contains)
+
+      // Node 0 rejected deploys
+      deploys         = deploysNode1 ++ deploysNode2
+      rejected        = blockMerged.body.rejectedDeploys.map(_.sig)
+      rejectedDeploys = deploys.values.map(_.sig).map(rejected.contains)
+
+      res = ListMap(
+        "Merged block    " -> blockMerged,
+        "All blocks      " -> node0Blocks,
+        "Parents         " -> mergedParents,
+        "Deploys         " -> deploys.keys,
+        "Rejected deploys" -> rejectedDeploys
+      )
+      _ = println(s"RESULT:\n${res.mkString("\n")}")
+    } yield ()
+  }
+
+  it should "reject deploys built on top of rejected deploy (separate blocks)" in effectTest {
+    TestNode.networkEff(genesis, networkSize = 3).use { nodes =>
+      testCase(nodes) { deploys =>
+        // Set and read in separate blocks
+        deploys.traverse(nodes(2).addBlock(_))
+      }
+    }
+  }
+
+  it should "reject deploys built on top of rejected deploy (same block)" in effectTest {
+    TestNode.networkEff(genesis, networkSize = 3).use { nodes =>
+      testCase(nodes) { deploys =>
+        // Set and read in the same block
+        nodes(2).addBlock(deploys: _*).map(List(_))
```

Comment:
When deploys are in the same block, then _set_ and _read_ are rejected.
```
Deploys          -> Set(N1_nila, N1_set, N1_nilb, N2_set, N2_read)
Rejected deploys -> List(false, false, false, true, true)
```

Reply:
Here is the DAG
  ```
/**
    *          *
    *         / \
    * set -> *   * <- set & read
    *         \ /
    *          *
    *        1 0 2
    */ 
``` 
Here it depends on ordering of block hashes which one picked as a main state, so either `set` from left is rejected - or `set` and `read` from right.

Reply:
@nzpr If dependency rejection is not implemented why is in this case _read_ deploy (N2_read) rejected?

Reply:
@tgrospic when channel is conflicting - all deploys touching this channel on are rejected. To test dependency you need to have some non conflicting channel which two or more deploys share. So it’s not a dependency that is tested here - it’s just rejection. I think I found a bug that prevents read deploy from rejecting in multi block settings.